Feature/logfilters 17

kdeme opened the pull request:
Here is a first proposal for #17. I say first, as I feel there will be iterations ;)

Some items I came across during development:
- The documentation mentions that the topics are comma- or space-separated lists (both in context of compile time option as for logScope or log statement).
 Currently this was not the case. I added comma in both contexts. If this is not wanted, we should update the documentation.
- chronicles_enabled_topics and chronicles_required_topics is probably something you will not use together. However they are accepted together, basically creating non working behaviour for both if you do so. This was adjusted somewhat with this PR but it all still depends on what is wanted in this situation. Perhaps just not allow both?
- In chronicles_enabled_topics the log level is allowed to be the enum values or its corresponding integer values.
  - I question if these integer values are so useful. If they would map on the levels of syslog/kernel or some other well known numbers perhaps, but as is?
  - I could have added the digit checking also in the handleEnumOption proc directly, but that would have as (I believe) unwanted side-effect that other enum options could also be set with their int values.

In general, I also believe the compile time commit could be more clean, however I think that will need a bigger code overhaul. Hence, I await a second opinion. :)

zah commented:
Thanks for working on this!

Nice catches regarding the documentation issues. I like your fix about the commas and the suggestion to just signal an error when `chronicles_enabled_topics` and `chronicles_required_topics` are used together.

The support for numeric log levels is there to let programmers save some typing if they know what they are doing. I think the numbers should just match the values of the `LogLevel` enum.

Most of my comments are minor nitpicks, but the `enabledTopicsMatch` remark seems like a potential issue to me. Please take a look at it.

zah commented on chronicles/topics_registry.nim:
```diff
@@ -7,31 +8,37 @@ type
     Required,
     Disabled
 
+  Topic* = object
```
Ah, silly me, this doesn't capture all the cases properly.

kdeme commented on chronicles/options.nim:
```diff
@@ -88,6 +88,10 @@ type
   Configuration* = object
     streams*: seq[StreamSpec]
 
+  Topic* = object
```
Actually, not just for clarity, also to make it build ;). Ambiguous identifier right now, sorry should have noticed this at testing.

kdeme commented on chronicles.nim:
```diff
@@ -180,6 +184,7 @@ macro logIMPL(lineInfo: static InstInfo,
   # This is the compile-time topic filtering code, which has a similar
   # logic to the generated run-time filtering code:
   var enabledTopicsMatch = enabledTopics.len == 0
+  var enabledTopicsLevel = enabledLogLevel
```
Yes, you are right. I'll fix this.

kdeme commented:
Added commit that should resolve the issues. I've tested more extensively (there were much more cases to cover then I initially anticipated). I guess we really need #20, I do not want to run all this testing manually again ;). Perhaps a good next task for me.

During testing I noticed another issue when setting the log level to NONE. For example:`-d:chronicles_enabled_topics:main:NONE`.
The existing `handleEnumOption` proc would set this to default value 0 which is `DEBUG` for the `LogLevel` enum. I have changed to location of NONE in the enum, but of course this means the numbers go all up one. If this is not wanted I can also change the `handleEnumOption` proc to allow for an optional `defaultOff` parameter which is default set to 0.

Next to that I've moved the actually checking on `enabledTopicsMatch` above the line number code as improvement.

Was there something you wanted changed still on the `Topic` object?

I did not yet implement the error signal of `chronicles_enabled_topics` vs `chronicles_required_topics`. OK if I do this in another PR?
In the current version they work, the required topics have precedence over enabled.

zah commented:
OK, great, I'm quite happy with the current state of the code.

It seems to me that this was more substantial work than the recently awarded #17, so we can set the payout to 400 USD. @kdeme, can you send us an Ethereum address suitable for receiving ETH, SNT or DAI?

Regarding #20, I agree it would be quite an appropriate next task.

kdeme commented:
Great, Thanks.

ETH or SNT is perfect. My address: 0x1aD0BDF0970e511BFa2731AB59428c6d89ca71bb

pablanopete commented:
@kdeme will be following up as soon as I get confirmation from @zah!

zah commented:
@pablanopete, please go on. This work has been already merged, so this means we have accepted it.

kdeme commented:
@pablanopete, I have not yet received the bounty on the provided address, could you have a look please?

zah commented:
@kdeme, I'm very sorry for the delay. I've personally contacted the right people and made sure that the bounty was paid today:

https://etherscan.io/tx/0xcedfcf2604abd7b56c7ac1fec615c74a0e25b33090515b167e26e157b688c96f
